src/Process/Signal.php

```diff
@@ -13,113 +13,97 @@ class Signal implements SignalInterface
     use Defensive\AwareTrait;
     use Process\Signal\Information\Factory\AwareTrait;
     use Process\Pool\Logger\AwareTrait;
-    protected $_waitCount       = 0;
-    protected $_signalHandlers  = [];
-    protected $_bufferedSignals = [];
+    protected $signalHandlers = [];
+    protected $bufferedSignals = [];
 
     public function addSignalHandler(int $signalNumber, HandlerInterface $signalHandler): SignalInterface
     {
-        $this->incrementWaitCount();
         pcntl_async_signals(true);
-        $this->_signalHandlers[$signalNumber] = $signalHandler;
+        $this->signalHandlers[$signalNumber] = $signalHandler;
         pcntl_signal($signalNumber, [$this, 'handleSignal']);
-        $this->decrementWaitCount();
 
         return $this;
     }
 
-    public function incrementWaitCount(): SignalInterface
+    public function processBufferedSignals(): SignalInterface
     {
-        ++$this->_waitCount;
-
-        return $this;
-    }
-
-    public function decrementWaitCount(): SignalInterface
-    {
-        $this->block();
-        if ($this->_waitCount === 1) {
-            $this->_processBufferedSignals();
+        foreach ($this->bufferedSignals as $position => $information) {
+            unset($this->bufferedSignals[$position]);
+            $this->processSignalInformation($information);
```

Comment:
Yeah, since the VM prevents being completely event driven I can rearrange the order of execution of these two lines and wrap it in a try catch and re-throw. If a Handler attempted processing a signal and encountered some sort of unhandled fault then Signal has to remove it. Good catch.